New plugin: tangorin

winterthediplomat opened the pull request:
This is a (possible) implementation for the `tangorin` plugin that is already present in Descartes but not in Gauss.

I didn't write lots of Rust until now, so if you spot something that is not idiomatic, or could be done better, or just mistakes I overlooked, I'll be more than happy to fix them.

Fixes #2

winterthediplomat committed:
plugins/tangorin: implement romaji and kana retrieval

winterthediplomat committed:
plugins/tangorin: retrieve the first kanji in the list in a different function

winterthediplomat committed:
plugins/tangorin: add meaning in the generated reply, fix test names

winterthediplomat committed:
plugins/tangorin: implement the 'info' retrieval

winterthediplomat committed:
plugins/tangorin: remove duplicated code, rename deeper_text to inner_text

winterthediplomat committed:
plugins/tangorin: simplify capturing, rename capturing method

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+impl Tangorin {
+    fn grep_kanji(&self, msg: &str) -> Option<String> {
+        match RE.captures(msg) {
+            Some(captures) => {
```
`Some(captures) => captures.at(1).map(|e| e.to_owned())`

winterthediplomat replied:
done

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+
+    fn retrieve_from_selector(&self, doc: &kuchiki::NodeRef, selector: &str) -> Option<String> {
+        if let Some(match_) = doc.select(selector).unwrap().next() {
+            let node    = match_.as_node().first_child().unwrap();
```
check the indentation here, it looks quite weird

winterthediplomat replied:
ssh on spotty connections is suffering and makes me miss this kind of errors. fixed.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+        if let Some(match_) = doc.select(selector).unwrap().next() {
+            let node    = match_.as_node().first_child().unwrap();
+            let borrowed_text   = node.as_text().unwrap().borrow();
+            let mut retrieved_text = borrowed_text.clone();
```
I think this doesn't need to be mutable as you should be able to do everything in a single line, right?

winterthediplomat replied:
now this not mutable anymore. Originally, I had some problems with `borrowed_text` lifetimes: to fix it I put each change to its own line, but didn't recompose things back.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+            
+            Some(retrieved_text)
+        }
+        else{
```
I don't like much this line. What about an early return on the very top of the function?

winterthediplomat replied:
refactored the whole thing, so early returns are not necessary anymore.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+        if let Some(match_) = doc.select("span[class=eng]").unwrap().next() {
+            let node = match_.as_node(); 
+            let children = node.children();
+            let mut meaning = "".to_string();
```
`String::new()` in place of `"".to_string()`

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+            let node = match_.as_node(); 
+            let children = node.children();
+            let mut meaning = "".to_string();
+            for child_ in children {
```
What about just `child`?

winterthediplomat replied:
old habits are hard to kill. fixed now.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+            let children = node.children();
+            let mut meaning = "".to_string();
+            for child_ in children {
+                if let Some(text_) = child_.as_text() {
```
`text`?

winterthediplomat replied:
old habits, again.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+
+           Some(meaning)
+        }
+        else{
```
As before, maybe an early return would work better

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+                    None => None
+                }
+            }
+            else{
```
I don't find value in this `else`

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+                Some(result_str) => result_str.borrow(),
+                None => { return None }
+            };
+            Some(res.clone())
```
prefer `to_owned()` in place of `clone()`. Btw I think this block can be make shorter

winterthediplomat replied:
the code now uses `to_owned`.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+
+    fn tangorin(&self, server: &IrcServer, _: &Message, target: &str, msg: &str) -> io::Result<()> {
+        let url = match self.grep_kanji(msg) {
+            Some(kanji) => "http://tangorin.com/general/".to_string() + kanji.as_str(),
```
Let's use `format!` for this

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+        };
+
+        if let Ok(doc) = kuchiki::parse_html().from_http(&url) {
+            let romaji = match self.retrieve_from_selector(&doc, "rt"){
```
What if we make a macro to reduce this redundant lines?

winterthediplomat replied:
implemented a simple macro called `get_text!`. it definitely cuts the noise down.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+            let info : String = match self.retrieve_info(&doc) {
+                Some(retrieved) => {
+                    let sanitized = retrieved.replace("\u{2014}", "").replace(".", "").to_lowercase();
+                    (" (".to_string()+sanitized.as_str()+")").to_string()
```
`format!`

winterthediplomat replied:
i felt there was a better way to do that, but didn't look at the docs. fixed.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+                    let sanitized = retrieved.replace("\u{2014}", "").replace(".", "").to_lowercase();
+                    (" (".to_string()+sanitized.as_str()+")").to_string()
+                },
+                None => "".to_string()
```
`String::new`

winterthediplomat replied:
fixed.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+                None => "".to_string()
+            };
+          
+            return server.send_privmsg(target, &format!("[Tangorin] {} ({} - {}): {}{}", &*kanji, &*kana, &*romaji, &*meaning, &*info))
```
I'm pretty sure that filling a proper struct with these data and using the `ToString` trait would work great

winterthediplomat replied:
I don't really think so. Creating an useless struct just to use `.to_string` is not as readable as a `format!`, especially considering that the retrieved information is not shared with other plugins or stored in any way.

RoxasShadow replied:
At least let's try to format that string properly (splitting the vars in multiple lines just work fine) as right now that's. `struct` as very lightweight btw :)

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,214 @@
+        assert_eq!("PRIVMSG test :[Tangorin] 頑な (かたくな - katakuna): obstinate (usually written using kana alone)\r\n",
+                   &*get_server_value(&server));
+    }
+
```
useless line here, one is enough

winterthediplomat replied:
should be fixed

winterthediplomat committed:
plugins/tangorin: fix typos and suggestions

winterthediplomat committed:
plugins/tangorin: replace redundant code with get_text! macro

winterthediplomat committed:
plugins/tangorin: refactor text retrieval code

in `retrieve_*` function, use matches instead
of `if let ... else` to make code smaller
and more readable

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+            Some(match_) => {
+                let node = match_.as_node().first_child().unwrap();
+                let borrowed_text = node.as_text().unwrap().borrow();
+                Some(borrowed_text.clone().trim().to_string())
```
`to_owned()` is better than `clone()` as it is an its specialisation

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+    static ref RE: Regex = Regex::new(r"!tangorin (\S+)").unwrap();
+}
+
+macro_rules! get_text {
```
`try_option` to be more similar to `try`?

winterthediplomat replied:
done.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+
+    fn retrieve_meaning(&self, doc: &kuchiki::NodeRef) -> Option<String> {
+        match doc.select("span[class=eng]").unwrap().next() {
+            None => None,
```
Can you think a way to avoid this `None => None` in many places? like an early return or so

winterthediplomat replied:
as discussed privately, I've switched to `Options.map` when the conversion is easy and meaningful.

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+                    if let Some(text) = child.as_text() {
+                        meaning.push_str(text.borrow().as_str());
+                    }
+                    else{
```
put a space after `else` please

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+                let following_siblings = node.following_siblings();
+
+                match following_siblings.select("i[class=d-info]") {
+                    Err(()) => None,
```
`Err(_)`?

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+
+                match following_siblings.select("i[class=d-info]") {
+                    Err(()) => None,
+                    Ok(mut info_sibs) => {
```
you can avoid `{...}` as using direct `=> match` is enough

winterthediplomat replied:
good catch

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+                match following_siblings.select("i[class=d-info]") {
+                    Err(()) => None,
+                    Ok(mut info_sibs) => {
+                        match info_sibs.next() {
```
`info_sibs.next().map(|n| ...)`

winterthediplomat replied:
done

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,195 @@
+        match root.first_child(){
+            None => None,
+            Some(match_) => {
+                match match_.as_text() {
```
`match_.as_text().map(|n| ...)`

winterthediplomat replied:
done

winterthediplomat committed:
plugins/tangorin: implement second batch of fixes

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -30,36 +30,30 @@ impl Tangorin {
-        match doc.select(selector).unwrap().next() {
-            Some(match_) => {
+        doc.select(selector).unwrap().next().map(|match_| {
                 let node = match_.as_node().first_child().unwrap();
```
there's an indentation problem here maybe?

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -30,36 +30,30 @@ impl Tangorin {
+                if let Some(text) = child.as_text() {
+                    meaning.push_str(text.borrow().as_str());
+                }
+                else {
```
elsif?

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,186 @@
+
+    fn retrieve_from_selector(&self, doc: &kuchiki::NodeRef, selector: &str) -> Option<String> {
+        doc.select(selector).unwrap().next().map(|match_| {
+                let node = match_.as_node().first_child().unwrap();
```
indentation problem here

winterthediplomat replied:
forgot to fix it in the previous patch :V

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    fn inner_text(&self, root: &kuchiki::NodeRef) -> Option<String> {
+        match root.first_child(){
```
put a space before the `{`

winterthediplomat replied:
fixed

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,186 @@
+            let kana = try_option!(self.retrieve_from_selector(&doc, "rb"));
+            let kanji = try_option!(self.retrieve_from_selector(&doc, "span[class=writing]"));
+            let meaning = try_option!(self.retrieve_meaning(&doc));
+            let info : String = match self.retrieve_info(&doc) {
```
don't put any space before the `:` (i.e. `let info: String =`)

winterthediplomat replied:
ops :///:

RoxasShadow commented on src/plugins/tangorin.rs:
```diff
@@ -0,0 +1,186 @@
+            let kanji = try_option!(self.retrieve_from_selector(&doc, "span[class=writing]"));
+            let meaning = try_option!(self.retrieve_meaning(&doc));
+            let info : String = match self.retrieve_info(&doc) {
+                Some(retrieved) => format!(" ({})", retrieved.replace("\u{2014}", "").replace(".", "").to_lowercase()),
```
WDYT about `String::remove` here?

winterthediplomat replied:
unfortunately [String::remove](https://doc.rust-lang.org/std/string/struct.String.html#method.remove) does not do what we thought it does: 
`Removes a char from this String at a byte position and returns it.`.
String::replace is still the only method to do that.

RoxasShadow commented:
This is cool, just a couple of request changes and we can merge this – thanks a lot  👍

winterthediplomat committed:
plugins/tangorin: fix indentation and typos
